yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtilSuite.scala

```diff
@@ -245,4 +247,28 @@ class YarnSparkHadoopUtilSuite extends SparkFunSuite with Matchers with Logging
       System.clearProperty("SPARK_YARN_MODE")
     }
   }
+
+  test("Obtain tokens For HiveMetastore") {
+    val hadoopConf = new Configuration()
+    hadoopConf.set("hive.metastore.kerberos.principal", "bob")
+    // thrift picks up on port 0 and bails out, without trying to talk to endpoint
+    hadoopConf.set("hive.metastore.uris", "http://localhost:0")
+    val util = new YarnSparkHadoopUtil
+    val e = intercept[InvocationTargetException] {
+      val token = util.obtainTokenForHiveMetastoreInner(hadoopConf, "alice")
+      fail(s"Expected an exception, got the token $token")
+    }
+    val inner = e.getCause
+    if (inner == null) {
+      fail("No inner cause", e)
+    }
+    if (!inner.isInstanceOf[HiveException]) {
+      fail(s"Not a hive exception", inner)
+    }
+    // expect exception trapping code to unwind this hive-side exception
+    intercept[HiveException] {
+      val token = util.obtainTokenForHiveMetastore(hadoopConf)
+      fail(s"Expected an exception, got the token $token")
```

Comment:
again, better diagnostics. Now, if you look at `Assertions.intercept()`, it's clearly discarding the result of the inner closure. If it  was redone to take an error string on a mismatch, and prepend that to the `.toString()` value of anything returned which wasn't of type `Unit`, then we'd have something really good at reporting failures to raise exceptions. As it is, `intercept` is pretty lossy.